drivers/SmartThings/zigbee-thermostat/src/popp/common.lua

```diff
@@ -0,0 +1,222 @@
+local TemperatureAlarm = capabilities.temperatureAlarm
+local Switch = capabilities.switch
+local TemperatureMeasurement = capabilities.temperatureMeasurement
+local ThermostatHeatingSetpoint = capabilities.thermostatHeatingSetpoint
+
+local last_setpointTemp = nil
+local common = {}
+
+common.MIN_SETPOINT = 5
+common.MAX_SETPOINT = 30
+common.STORED_HEAT_MODE = "stored_heat_mode"
+
+common.THERMOSTAT_CLUSTER_ID = 0x0201
+common.MFG_CODE = 0x1246
+common.WINDOW_OPEN_FEATURE = nil
+
+common.THERMOSTAT_SETPOINT_CMD_ID = 0x40
+common.WINDOW_OPEN_DETECTION_ID = 0x4000
+common.WINDOW_OPEN_DETECTION_MAP = {
+  [0x00] = "cleared", -- // "quarantine" default
+  [0x01] = "cleared", -- // "closed" window is closed
+  [0x02] = "freeze", -- // "hold" window might be opened
+  [0x03] = "freeze", -- // "opened" window is opened
+  [0x04] = "freeze", -- // "opened_alarm" a closed window was opened externally (=alert)
+}
+common.EXTERNAL_OPEN_WINDOW_DETECTION_ID = 0x4003
+
+local SUPPORTED_MODES = {
+  ThermostatMode.thermostatMode.off.NAME,
+  ThermostatMode.thermostatMode.heat.NAME,
+  ThermostatMode.thermostatMode.eco.NAME
+}
+
+-- Helpers
+
+-- has member check function
+local has_member = function(haystack, needle)
+  for _, value in ipairs(haystack) do
+    if (value == needle) then
+      return true
+    end
+  end
+  return false
+end
+
+-- Turn switch 'on' if its state is 'off'
+local turn_switch_on = function(driver, device)
+  -- turn thermostat ventile on
+  if device:get_latest_state("main", Switch.ID, Switch.switch.NAME) == "off" then
+    common.switch_handler_factory('on')(driver, device)
+  end
+end
+
+-- Handlers
+
+-- Internal window open detection handler
+common.window_open_detection_handler = function(driver, device, value, zb_rx)
+  device:emit_event(TemperatureAlarm.temperatureAlarm(common.WINDOW_OPEN_DETECTION_MAP[value.value]))
+end
+
+-- Switch handler
+common.switch_handler_factory = function(switch_state)
+  return function(driver, device, cmd)
+    local get_cmd = switch_state or cmd.command
+    local external_window_open
+
+    if get_cmd == 'on' then
+      external_window_open = false
+    elseif get_cmd == 'off' then
+      external_window_open = true
+    end
+
+    device:send(cluster_base.write_manufacturer_specific_attribute(device, common.THERMOSTAT_CLUSTER_ID,
+      common.EXTERNAL_OPEN_WINDOW_DETECTION_ID,
+      common.MFG_CODE, data_types.Boolean, external_window_open))
+  end
+end
+
+-- Custom setpoint command handler
+common.setpoint_cmd_handler = function(driver, device, cmd)
+  local payload
+  local mode = cmd.args.mode
+
+  if has_member(SUPPORTED_MODES, mode) then
+
+    -- fetch last_setpointTemp
+    last_setpointTemp = device:get_field("last_setpointTemp")
+
+    if last_setpointTemp == nil then
+      last_setpointTemp = device:get_latest_state("main", ThermostatHeatingSetpoint.ID, ThermostatHeatingSetpoint.heatingSetpoint.NAME) or 21 -- thermostat default temperature
```

Comment:
is there ever a way that this `last_setpointTemp` value that you are storing will be different than the value returned by `get_latest_state`?